--- hobbit-4.1.2p1/client/hobbitclient-linux.sh.msgs	2005-10-09 16:11:33.000000000 -0400
+++ hobbit-4.1.2p1/client/hobbitclient-linux.sh	2006-01-24 04:45:58.632732032 -0500
@@ -37,6 +37,7 @@
 nohup sh -c "vmstat 300 2 1>$BBTMP/hobbit_vmstat.$$ 2>&1; mv $BBTMP/hobbit_vmstat.$$ $BBTMP/hobbit_vmstat" </dev/null >/dev/null 2>&1 &
 sleep 5
 if test -f $BBTMP/hobbit_vmstat; then echo "[vmstat]"; cat $BBTMP/hobbit_vmstat; rm -f $BBTMP/hobbit_vmstat; fi
+if [ -r $BBTMP/hobbit_msgs ]; then echo "[msgs]";cat $BBTMP/hobbit_msgs;fi
 
 exit
 
