--- branches/4.2.4/configure.client	2009/02/22 17:06:20	6161
+++ branches/4.2.4/configure.client	2009/02/22 20:40:23	6162
@@ -62,6 +62,9 @@
 fi
 echo ""; echo ""
 
+MAKE="$MAKE -s" . ./build/clock-gettime-librt.sh
+echo ""; echo ""
+
 echo "What userid will be running Xymon [xymon] ?"
 if test -z "$BBUSER"
 then
@@ -164,6 +167,7 @@
 echo "BBHOSTIP = $BBHOSTIP"              >>Makefile
 echo "# Large File Support settings"     >>Makefile
 echo "LFSDEF = $LFS"                     >>Makefile
+echo "LIBRTDEF = $LIBRTDEF"              >>Makefile
 
 echo "" >>Makefile
 if test -r build/Makefile.`uname -s`

